### PR TITLE
Basics

### PR DESCRIPTION
**Pull request by maximzubkov**
Create `ContrasiveDataset` and `TextDataset`, add base `LightningDataModule` and `LSTM` model compitable with dataset processed with `gorshochek`

### CODE REVIEW

File: scripts/download_poj.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
```
Comment (SpirinEgor) [inline scripts/download_poj.sh:1]: Вообще такие вещи можно прекрасно делать в `prepare_data` в `LigthningDataModule`. По крайней мере он для этого и создавался. Тогда у тебя при запуске модели будет проверяться наличие датасета, а если его нет, то скачиваться.

Так же есть `setup`. Там можно организовать построение словаря. Что если его нет, то запустить yttm

File: configs/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from .lstm_config import default_config, test_config, LSTMConfig # noqa
```
[the remark this answers is not in the document]
Comment (maximzubkov) [reply configs/__init__.py:1]: Да, так как импортируется, но не используется

File: models/encoders/lstm.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from configs import LSTMConfig
+
+
+class LSTMModel(nn.Module):
+    def __init__(self, config: LSTMConfig):
+        super().__init__()
+        self.num_classes = config.output_size
+        self.hidden_size = config.hidden_size
+        self.dropout = nn.Dropout(config.dropout)
+        self.embeddings = nn.Embedding(config.vocab_size, config.embedding_size, padding_idx=0)
```
[the remark this answers is not in the document]
Comment (maximzubkov) [reply models/encoders/lstm.py:13]: В `yttm` можно задать эти значения через аргументы, дефолтные занчения вот:
 ```
  --pad_id INTEGER      Padding token id.  [default: 0]
  --unk_id INTEGER      Unknown token id.  [default: 1]
  --bos_id INTEGER      'Begin of sentence' token id.  [default: 2]
  --eos_id INTEGER      'End of sentence' token id.  [default: 3]
  ```

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+mypy==0.800
+numpy==1.20.0
+pytorch-lightning==1.1.6
+pytorch-lightning-bolts==0.3.0
+torch==1.7.1
+tqdm==4.56.0
+wandb==0.10.15
+pytest==6.2.2
+Cython==0.29.21
+youtokentome
```
[the remark this answers is not in the document]
Comment (maximzubkov) [reply requirements.txt:11]: Так как это же не дефолтный `yttm`, а форк Ярослава. более корректно, наверное, написать скрпит, который будет создавать папку `build` и там ставить `yttm` Ярослава, а из `reqirements` его убрать

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+torchvision~=0.8.2
```
[the remark this answers is not in the document]
Comment (maximzubkov) [reply requirements.txt:12]: Я тестирую обертку для датасета при помощи датасета `MNIST` из `torchvision`

File: configs/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SpirinEgor) [reply configs/__init__.py:1]: А если в __all__ добавить?

File: models/encoders/lstm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SpirinEgor) [reply models/encoders/lstm.py:13]: Понял. Можно обезопасить себя как-нибудь, например, сделать что-то вроде `yttm.get_pad_idx`.

File: requirements.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (SpirinEgor) [reply requirements.txt:11]: Ну вот да. Потому что так может установиться оригинальный yttm

File: dataset/base_data_module.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        else:
+            raise NotImplementedError("Non-text datasets are currently not available")
+
+        self.dataset = ContrastiveDataset(clf_dataset=self.clf_dataset)
+
+    def setup(self, stage: str = None):
+        len_train = int(0.6 * len(self.dataset))
+        len_test = int(0.2 * len(self.dataset))
+        len_val = len(self.dataset) - len_test - len_train
+        self.train, self.val, self.test = random_split(self.dataset, [len_train, len_val, len_test])
```
Comment (SpirinEgor) [inline dataset/base_data_module.py:40]: Should you fix here random seed?
It seems unnecessary since this is a torch function. But it's better to recheck.
Also, what behavior you expected if a dataset is already split into holdouts?

File: dataset/base_data_module.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        # batch contains a list of tuples of structure (sequence, target)
+        a = pad_sequence([item["a_encoding"].squeeze() for item in batch], batch_first=True)
+        b = pad_sequence([item["b_encoding"].squeeze() for item in batch], batch_first=True)
+        return (a, b), None
+
+    def transfer_batch_to_device(self, batch: Any, device: torch.device) -> Any:
+        (a, b), label = batch
+        a = a.to(device)
+        b = b.to(device)
+        return (a, b), label
```
Comment (SpirinEgor) [inline dataset/base_data_module.py:62]: The label should be on the device too. Especially if your loss calculating using this label.

File: dataset/text_dataset.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        _, base_dirs, _ = next(walk(self.dataset_path))
+        base_dirs_paths = map(lambda file_: join(self.dataset_path, file_), base_dirs)
+        idx = 0
+        for base_dir_path in base_dirs_paths:
+            _, _, dir_files = next(walk(base_dir_path))
+            dir_files_paths = map(lambda file_: join(base_dir_path, file_), dir_files)
+            dir_files_paths = filter(self._in_c_family, dir_files_paths)
+            for file in dir_files_paths:
+                self.idx2file[idx] = file
+                idx += 1
```
Comment (SpirinEgor) [inline dataset/text_dataset.py:33]: I think the same logic would be reused in other datasets. So you can move it in utils or something like that.

File: models/self_supervised/moco.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        encoder_momentum: float = 0.999,
+        softmax_temperature: float = 0.07,
+        learning_rate: float = 0.03,
+        momentum: float = 0.9,
+        weight_decay: float = 1e-4,
+        batch_size: int = 256,
+        use_mlp: bool = False,
+        num_workers: int = 8,
+        *args,
+        **kwargs
```
Comment (SpirinEgor) [inline models/self_supervised/moco.py:25]: Do you really need these two fields?

File: preprocess/tokenize.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                    with open(file_path, "r", encoding="utf8", errors='ignore') as file_:
+                        text = file_.read() + "\n"
+                        with open(buffer_path, "a") as buffer_:
+                            buffer_.write(text)
+
+    tokenizer_config = test_tokenizer_config if is_test else default_tokenizer_config
+    tokenizer = yttm.BPE.train(
+        data="text.yttm",
+        model=model_path,
+        **tokenizer_config.__dict__
```
Comment (SpirinEgor) [inline preprocess/tokenize.py:33]: Don't use direct call. There are a special method for converting data class into the dict
https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

File: train.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        filename="{epoch:02d}-{val_loss:.4f}",
+        period=3,
+        save_top_k=-1,
+    )
+    # use gpu if it exists
+    gpu = 1 if torch.cuda.is_available() else None
+    # define learning rate logger
+    lr_logger = LearningRateMonitor("step")
+    trainer = Trainer(
+        max_epochs=10,
```
Comment (SpirinEgor) [inline train.py:56]: Should this be moved into hyperparameters?

File: train.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        deterministic=True,
```
Comment (SpirinEgor) [inline train.py:57]: That may slow the training? Do you need this since you already fix the seed?

File: train.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        check_val_every_n_epoch=1,
+        log_every_n_steps=50,
```
Comment (SpirinEgor) [inline train.py:59]: 50? Isn't that too much? How long 1 step take?

**Review by SpirinEgor — CHANGES_REQUESTED** [review]

File: dataset/base_data_module.py (hunk 6 above)
Comment (maximzubkov) [reply dataset/base_data_module.py]: There is no seed arg for `random_split`, I moved the logic of splitting the dataset to the `load_data` function

**Review by SpirinEgor — APPROVED** [review]
